cmd/limactl/volume.go

```diff
@@ -0,0 +1,235 @@
+	return volumeListCommand
+}
+
+func volumeListAction(cmd *cobra.Command, args []string) error {
+	allVolumes, err := store.Volumes()
+	if err != nil {
+		return err
+	}
+
+	w := tabwriter.NewWriter(cmd.OutOrStdout(), 4, 8, 4, ' ', 0)
+	fmt.Fprintln(w, "NAME\tSIZE\tDIR\tIN USE BY")
+
+	if len(allVolumes) == 0 {
+		logrus.Warn("No volume found. Run `limactl volume create VOLUME` to create a volume.")
+	}
+
+	for _, volName := range allVolumes {
+		vol, err := store.InspectVolume(volName)
+		if err != nil {
+			logrus.WithError(err).Errorf("volume %q does not exist?", volName)
+			continue
+		}
+		fmt.Fprintf(w, "%s\t%s\t%s\t%s\n", vol.Name, units.BytesSize(float64(vol.Size)), vol.Dir, vol.Instance)
+	}
+
+	return w.Flush()
+}
+
+func newVolumeDeleteCommand() *cobra.Command {
+	var volumeDeleteCommand = &cobra.Command{
+		Use: "delete VOLUME [VOLUME, ...]",
+		Example: `
+To delete a volume:
+$ limactl volume delete VOLUME
+
+To delete multiple volumes:
+$ limactl volume delete VOLUME1 VOLUME2 ...
+`,
+		Aliases: []string{"remove", "rm"},
+		Short:   "Delete one or more Lima volumes",
+		Args:    cobra.MinimumNArgs(1),
+		RunE:    volumeDeleteAction,
+	}
+	return volumeDeleteCommand
+}
+
+func volumeDeleteAction(cmd *cobra.Command, args []string) error {
+	for _, volName := range args {
+		vol, err := store.InspectVolume(volName)
+		if err != nil {
+			if errors.Is(err, fs.ErrNotExist) {
+				logrus.Warnf("Ignoring non-existent volume %q", volName)
+				continue
+			}
+			return err
+		}
+		if err := deleteVolume(vol); err != nil {
+			return fmt.Errorf("failed to delete volume %q: %w", volName, err)
+		}
+		logrus.Infof("Deleted %q (%q)", volName, vol.Dir)
+	}
+	return nil
+}
+
+func deleteVolume(vol *store.Volume) error {
+	if err := os.RemoveAll(vol.Dir); err != nil {
+		return fmt.Errorf("failed to remove %q: %w", vol.Dir, err)
+	}
+	return nil
+}
+
+func newVolumeUnlockCommand() *cobra.Command {
+	var volumeUnlockCommand = &cobra.Command{
+		Use: "unlock VOLUME [VOLUME, ...]",
+		Example: `
+Emergency recovery! If an instance is force stopped, it may leave a volume locked while not actually using it.
+
+To unlock a volume:
+$ limactl volume unlock VOLUME
+
+To unlock multiple volumes:
+$ limactl volume unlock VOLUME1 VOLUME2 ...
+`,
+		Short: "Unlock one or more Lima volumes",
+		Args:  cobra.MinimumNArgs(1),
+		RunE:  volumeUnlockAction,
+	}
+	return volumeUnlockCommand
+}
+
+func volumeUnlockAction(cmd *cobra.Command, args []string) error {
+	for _, volName := range args {
+		vol, err := store.InspectVolume(volName)
+		if err != nil {
+			if errors.Is(err, fs.ErrNotExist) {
+				logrus.Warnf("Ignoring non-existent volume %q", volName)
+				continue
+			}
+			return err
+		}
+		if vol.Instance == "" {
+			logrus.Warnf("Ignoring unlocked volume %q", volName)
+			continue
+		}
+		inst, err := store.Inspect(vol.Instance)
+		if err != nil {
+			logrus.Warnf("Could not inspect instance %q using volume %q", vol.Instance, volName)
+			continue
+		}
```

Comment:
This is not correct. When `store.Inspect()` returns an `err`, it means the instance does not exist. In that case we need to skip the check if the instance is running and go straight to unlocking the volume.

This can happen e.g. when you `limactl delete -f INSTANCE`.

I would argue that force-deleting an instance should still unlock the volumes too, but the logic here should be fixed anyways, in case there are other ways to end up with a deleted instance that leaves the symlinks behind.

If there are any errors besides the instance not existing, then they are returned in `inst.Errors`. If that slice is non-empty, then the errors should be reported here, and the unlock should be skipped.

Reply:
> When store.Inspect() returns an err, it means the instance does not exist. In that case we need to skip the check if the instance is running and go straight to unlocking the volume.

Thank you for fixing this!

> I would argue that force-deleting an instance should still unlock the volumes too, but the logic here should be fixed anyways, in case there are other ways to end up with a deleted instance that leaves the symlinks behind.

Please edit `stopInstanceForcibly` in `cmd/limactl/stop.go` and unlock all disks locked by the instance being killed!